File: Java/section1.1/beads.java
/*
ID: jameszh1
LANG: JAVA
TASK: beads
*/

import java.io.*;
public class beads {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("beads.in"));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("beads.out")));
        int numOfBeads = Integer.parseInt(br.readLine());
        String necklace = br.readLine();
//        int numOfBeads = 29;
//        String necklace = "wwwbbrwrbrbrrbrbrwrwwrbwrwrrb";
        necklace = necklace + necklace;
        int[][] left = new int[necklace.length() + 1][2];
        int[][] right = new int[necklace.length() + 1][2];
        left[0][0] = left[0][1] = 0;

        for (int i = 1; i <= necklace.length(); i++) {
            if (necklace.charAt(i - 1) == 'r') {
                left[i][0] = left[i - 1][0] + 1;
                left[i][1] = 0;
            } else if (necklace.charAt(i - 1) == 'b') {
                left[i][1] = left[i - 1][1] + 1;
                left[i][0] = 0;
            } else {
                left[i][0] = left[i - 1][0] + 1;
                left[i][1] = left[i - 1][1] + 1;
            }
        }

        right[necklace.length()][0] = right[necklace.length()][1] = 0;
        for (int i=necklace.length() - 1; i >= 0; i--){
            if (necklace.charAt(i) == 'r'){
                right[i][0] = right[i + 1][0] + 1;
                right[i][1] = 0;
            } else if (necklace.charAt(i) == 'b'){
                right[i][1] = right[i + 1][1] + 1;
                right[i][0] = 0;
            } else {
                right[i][0] = right[i + 1][0] + 1;
                right[i][1] = right[i + 1][1] + 1;
            }
        }

        int m = 0;
        for (int i = 0; i < necklace.length(); i++) {
            m = Math.max(m, Math.max(left[i][0], left[i][1]) + Math.max(right[i][0], right[i][1]));
        }
        m = Math.min(m, numOfBeads);
       // System.out.print(m);
        pw.println(m);
        pw.close();
        br.close();
    }

}

File: Python/chapter1/section2/beads.py
"""
ID: jameszh1
LANG: PYTHON2
TASK: beads
"""
fin = open ('beads.in', 'r')
fout = open ('beads.out', 'w')
number_of_bead = int(fin.readline().strip())
necklace = fin.readline().strip()
necklace = necklace + necklace
left = [([0 for row in range(2)]) for row in range(800)]
right = [([0 for col in range(2)]) for row in range(800)]

for i in range(1, 2 * number_of_bead + 1, 1):
    if (necklace[i - 1] == 'r'):
        left[i][0] = left[i - 1][0] + 1
        left[i][1] = 0
    elif (necklace[i - 1] == 'b'):
        left[i][1] = left[i - 1][1] + 1
        left[i][0] = 0
    else:
        left[i][0] = left[i - 1][0] + 1
        left[i][1] = left[i - 1][1] + 1

for i in range(2 * number_of_bead - 1, -1, -1):
    if (necklace[i] == 'r'):
        right[i][0] = right[i + 1][0] + 1
        right[i][1] = 0
    elif (necklace[i] == 'b'):
        right[i][1] = right[i + 1][1] + 1
        right[i][0] = 0
    else:
        right[i][0] = right[i + 1][0] + 1
        right[i][1] = right[i + 1][1] + 1
m = 0
for i in range(0, 2 * number_of_bead):
    m = max(m, max(left[i][0], left[i][1]) + max(right[i][0], right[i][1]))
m = min(m, number_of_bead)
fout.write (str(m) + '\n')
fout.close()